### PR TITLE
Add shard configuration example

### PR DESCRIPTION
**Pull request by miguelportilla**

### CODE REVIEW

File: doc/rippled-example.cfg [hunk 1]
```diff
@@ -1076,6 +1076,13 @@ file_size_mult=2
 online_delete=2000
 advisory_delete=0
 
+# This is the persistent datastore for shards. Helpful information can be
+# found here: https://ripple.com/build/history-sharding
```
Comment (bachase) [inline doc/rippled-example.cfg]: Consider adding a suggestion to uncomment the lines below to enable history sharding, maybe something like

```
# The commented lines below show how to configure the persistent datastore for shards.
# Consider uncommenting to enable the feature and contribute to the health of the Ripple network.
# Helpful information on the sharding can be found here: https://ripple.com/build/history-sharding
```

File: doc/rippled-example.cfg [hunk 2]
```diff
@@ -1076,6 +1076,13 @@ file_size_mult=2
...
+#[shard_db]
+#type=NuDB
+#path=/var/lib/rippled/db/shards/nudb
+#max_size_gb=500
```
Comment (MarkusTeufelberger) [inline doc/rippled-example.cfg]: How about enabling a 20-50GB shard_db in the config below?

File: doc/rippled-example.cfg (hunk 1 above)
Comment (miguelportilla) [reply doc/rippled-example.cfg]: @bachase Thanks. As per Arthur's suggestion, uncommented by default.

File: doc/rippled-example.cfg (hunk 2 above)
Comment (miguelportilla) [reply doc/rippled-example.cfg]: Updated the default to be enabled.

**Comment by scottschurr** [comment]
Unfortunately I'm going to expand the scope (at least of the thinking) for this pull request.

My question is, who is the intended audience of the rippled-example.cfg?

What I believe we've seen is that there are a number of relatively casual folks who are interested in running a rippled node for whatever reason.  They put together a system, either AWS or inexpensive local hardware, that seems reasonable to them.  Then they fire up the node using this rippled-example.cfg either verbatim or almost verbatim.  Then the node doesn't work well.

This is a problem.  It's discouraging for the folks who go to the effort to assemble a node.  It requires support work from rippled developers and the Customer Success team.  It looks bad and it's inefficient.

My take is that this proposed change to the rippled-example.cfg makes the problem worse.  If you look at the entire file you'll see that the `[node_db]` is set up for RocksDB.  So we clearly expect them to be running a spinning disk with lots of RAM.  Now we're adding the expectation that our user should be keeping shards.  And those shards are expected to be kept on an SSD.  The `[node_db]` and the `[shard_db]` stanzas are in direct opposition to each other.  With this change we can almost guarantee that the rippled-example.cfg file cannot be used by anyone without careful study and editing.

What's the solution?

I'm not sure I have a solution, but here's a thought.  Let's offer more than one example config file.  I'm not exactly sure how to divide up our potential customers, but maybe something along the line of:

- The cheapest viable AWS node that can stay synced to the network.  Only enough guts to stay synced and service one client.  Maybe, along with describing the AWS node, we describe local hardware that would support the same configuration (as long as it is running nothing besides rippled and has a good internet connection).  Would we need both SSD and spinning disk versions of this?

- A moderate node that can serve several clients and keep some shards and a couple of months of history.  Maybe this one could optionally be configured as a validator.  Would we need both SSD and spinning disk versions of this?

- An industrial strength full history node.  And describe the hardware that allows it to work.

But I shouldn't be making the call on the types of nodes we document.  That should come from the Product team and the Customer Success team.

If we supply different example configuration files for different node types, then we probably ought to move the documentation out of the configuration files.  We don't want to be maintaining identical documentation in multiple files in parallel.

At any rate, I've just described a big task.

The smaller task is to get an example for setting up shards in our rippled-example.cfg.  If we do that we should acknowledge that the file will be targeted at moderate-sized nodes.  We should decide whether that moderate-sized node is running on a spinning disk or on SSD, and be consistent about that throughout the entire file.  Let's make the example that we have work well for at least one specific type of node.  And we should document the minimum requirements of that node.

**Comment by MarkusTeufelberger** [comment]
RocksDB isn't even available on Windows afaik, `[node_size] medium` is not exactly small in resource usage, `advisory_delete 0` is the default value anyways... the example file has its issues, yes. Still it is just an example and it shouldn't be used without modifications anyways.

Making it (potentially) slightly broken could motivate people to actually configure their node correctly.

**Comment by miguelportilla** [comment]
@scottschurr @MarkusTeufelberger good points, thank you. Better documentation is really the best solution I can think of. A single file in the same directory that guides the operator on hardware and setup requirements for various rippled roles could be helpful.

**Comment by scottschurr** [comment]
As a compromise position, may I suggest that current rippled-example.cfg at least be self consistent?  Perhaps by using RocksDB for the shards database?

**Comment by miguelportilla** [comment]
@scottschurr NuDB is best suited for shards and we recommend using it (https://ripple.com/build/history-sharding/). NuDB uses fewer file handles per shard than RocksDB. RocksDB uses memory that scales with the size of data it stores, which may require excessive memory overhead with shards. RocksDB can sometimes outperform NuDB and therefore recommended as the default for the node store. Using RocksDB for the node store and NuDB for the shard store leverages the strengths of both technologies.

**Comment by scottschurr** [comment]
@miguelportilla, thanks for that response.  What you're saying makes perfect sense within its bounds.  In effect what you're saying is that a node that supports shards should be running on SSDs.  I'm good with that.  You're also saying that this example configuration file is intended to be used successfully on a wide variety of hardware, including non-SSD hardware.  I'm good with that too.

However by including the un-commented-out `[shard_db]` stanza, we are changing this particular example configuration file so it will take more work for new users to be successful with the file.  If we include the `[shard_db]` stanza, but comment it out, then the information is disseminated but we don't impair the example config file beyond its current usability.

There's also a potential downside to including the un-commented-out `[shard_db]` stanza.  With the stanza in place the first thing we're likely to tell new users who are having problems is to comment out the stanza.  This could be interpreted as reflecting badly on the usability and stability of the feature.  We want shards to be adopted.  Let's not put an obstacle in its way.

As long as we provide a single example config file, which is the situation today, my personal preference is that people need to tweak that example as little as possible in order to be successful.  Adding the un-commented-out `[shard_db]` stanza using `NuDB` takes us away from that direction.

I understand that the doc team is working on addressing concerns regarding the diversity of config file needs based on hardware and use cases.  My suggestion at this point is that a config file with an un-commented-out `[shard_db]` stanza should come along with that effort.

That said, this is probably less of a technical issue and more of a product and marketing issue.  We're impacting the usability of our example.  Someone from product of marketing should weigh in and declare whether they prefer wider usability, or including shards.

**Comment by miguelportilla** [comment]
@scottschurr @bachase @MarkusTeufelberger Thank you for the review and feedback.

**Comment by bachase** [comment]
In 1.0.0-b3 as d9be0de269712b79aedfb51973e4d9ef5f66b40c

**Pull request closed without merging**